### PR TITLE
feat(component-store): add patchState method

### LINKED ISSUE #2786: Add patchState method to the ComponentStore
Add `patchState` method to get rid of unnecessary `state` spreading and replace:

```typescript
// case 1
this.moviesStore.setState(state => ({
  ...state,
  movies,
}));

// case 2
this.moviesStore.setState(state => ({
  ...state,
  movies: [...state.movies, movie],
}));
```

with:

```typescript
// case 1
this.moviesStore.patchState({ movies });

// case 2
this.moviesStore.patchState(state => ({
  movies: [...state.movies, movie],
}));
```

## If accepted, I would be willing to submit a PR for this feature

[x] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

```typescript
// case 1
this.moviesStore.setState(state => ({
  ...state,
  movies,
}));

// case 2
this.moviesStore.setState(state => ({
  ...state,
  movies: [...state.movies, movie],
}));
```

Closes #2786

## What is the new behavior?

```typescript
// case 1
this.moviesStore.patchState({ movies });

// case 2
this.moviesStore.patchState(state => ({
  movies: [...state.movies, movie],
}));
```

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: modules/component-store/spec/component-store.spec.ts [hunk 1]
```diff
@@ -511,6 +541,45 @@ describe('Component Store', () => {
...
+            s: { ...INIT_STATE, value1: 'val1' },
+          })
+        );
+      })
+    );
+
+    it(
+      'with a value based on the previous state',
+      marbles((m) => {
+        componentStore.patchState(() => ({ value2: { foo: 'fooBar' } }));
```
Comment (timdeschryver) [inline modules/component-store/spec/component-store.spec.ts:572]: Should this test use the current state to build a new state?

e.g.

```suggestion
        componentStore.patchState((state) => ({ value2: { foo: state.value2.foo + "2" } }));
```

**Review by timdeschryver — APPROVED** [review]
It keeps surprising (in a good way) me that this code "is so simple" 👍

File: modules/component-store/spec/component-store.spec.ts (hunk 1 above)
Comment (markostanimirovic) [reply modules/component-store/spec/component-store.spec.ts:572]: sure 🙂

File: modules/component-store/src/component-store.ts [hunk 2]
```diff
@@ -156,6 +156,29 @@ export class ComponentStore<T extends object> implements OnDestroy {
...
+   * Patches the state with provided partial state.
+   *
+   * @param partialStateOrUpdaterFn a partial state or a partial updater
+   * function that accepts the state and returns the partial state.
+   * @throws Error if the state is not initialized.
+   */
+  patchState(
+    partialStateOrUpdaterFn: Partial<T> | ((state: T) => Partial<T>)
+  ): void {
+    if (!this.isInitialized) {
```
Comment (alex-okrushko) [inline modules/component-store/src/component-store.ts:169]: updater does this check, so it's not needed here.

File: modules/component-store/src/component-store.ts [hunk 3]
```diff
@@ -156,6 +156,29 @@ export class ComponentStore<T extends object> implements OnDestroy {
...
+      throw new Error(this.notInitializedErrorMessage);
+    }
+
+    const updaterFn = (state: T) => ({
```
Comment (alex-okrushko) [inline modules/component-store/src/component-store.ts:173]: a few suggestions:
1. We can use `this.setState` here
2. Let's extract and evaluate `patchedState` separately for readability purposes.

```suggestion
    this.setState((state) => {
      const patchedState = (typeof partialStateOrUpdaterFn === 'function') 
          ? partialStateOrUpdaterFn(state)
          : partialStateOrUpdaterFn;
      return {
      ...state,
      ...patchedState, // 👈 comes after `state` to override the properties
      };
    }
```

**Review by alex-okrushko — CHANGES_REQUESTED** [review]
Thanks for working on this 👍

File: modules/component-store/src/component-store.ts (hunk 2 above)
Comment (markostanimirovic) [reply modules/component-store/src/component-store.ts]: done

File: modules/component-store/src/component-store.ts (hunk 3 above)
Comment (markostanimirovic) [reply modules/component-store/src/component-store.ts]: done

**Review by brandonroberts — APPROVED** [review]

**Review by alex-okrushko — APPROVED** [review]

**Pull request merged**